File: Cpp/0786_k_th_smallest_prime_fraction.cpp
// https://leetcode.cn/problems/k-th-smallest-prime-fraction/

class Solution {
public:
    vector<int> kthSmallestPrimeFraction(vector<int>& arr, int k) {
        auto cmp = [&](const pair<int,int>& pair1, const pair<int,int>& pair2) {
            return arr[pair1.first] * arr[pair2.second] > arr[pair1.second] * arr[pair2.first];
        };

        int n = arr.size();
        priority_queue<pair<int,int>, vector<pair<int,int>>, decltype(cmp)> pq(cmp);
        for (int i = 0; i < n-1; ++i) {
            pq.push(make_pair(i, n-1));
        }

        for (int i = 0; i < k-1; ++i) {
            pair<int,int> idxs = pq.top();
            pq.pop();

            if (idxs.second - 1 > idxs.first) {
                pq.push(make_pair(idxs.first, idxs.second-1));
            }
        }
        vector<int> res = {arr[pq.top().first], arr[pq.top().second]};

        return res;
    }
};

File: Python3/0786_k_th_smallest_prime_fraction.py
# https://leetcode.cn/problems/k-th-smallest-prime-fraction/

class Solution:
    def kthSmallestPrimeFraction(self, arr, k):
        """
        Inputs:
            arr: list[int]
            k: int
        Outputs:
            res: list[int]
        """
        
        n = len(arr)
        pq = [] # priority queue
        for i in range(n-1):
            heapq.heappush(pq, [arr[i] / arr[n-1], i, n-1])

        for _ in range(k):
            frac, i, j = heapq.heappop(pq)

            if j-1 > i:
                heapq.heappush(pq, [arr[i] / arr[j-1], i, j-1])

        res = [arr[i], arr[j]]
        return res